Restructure and reimplement v-sync options

Geometror opened the pull request:
This PR is a complete reorginization of the v-sync related code for DisplayServers with the changes discussed here and on IRC. So it's no longer a reimplementation of PR #36862, altough @giarve's work was very helpful.

- Adds a v-sync mode setting which allows to choose between DISABLED, ON, ADAPTIVE and MAILBOX (fixes #36467, fixes godotengine/godot-proposals#1283)
	-> can be modified via project settings or via code at runtime [for Windows, Linux and OSX]
	-> fallback to VK_PRESENT_MODE_FIFO_KHR if the desired mode is not supported
	-> No engine restart is required for a change of the vsync-mode setting
- Removed the v-sync via compositor option as it was problematic (as suggested by @Calinou)

Geometror commented:
Rebased and partly restructured, 
vsync setting is now read in Main::setup2 + removed dead code in OS

What is the plan for vsync_via_compositor in 4.0? The code regarding vsync_via_compositor is currently dead in master.

Calinou commented:
> What is the plan for vsync_via_compositor in 4.0? The code regarding vsync_via_compositor is currently dead in master.

I think it would be better to remove it, as it doesn't work correctly for everyone. It causes the framerate to halve unexpectedly on some setups.

Geometror commented:
Ok, I removed the V-Sync via Compositor option entirely, now its ready for review :)

clayjohn commented:
Looks interesting, but is this not just a reimplementation of https://github.com/godotengine/godot/pull/36862? See this comment https://github.com/godotengine/godot/pull/36862#issuecomment-765469169 for reduz' response to that PR. 

I echo reduz' concern and I am not certain that just grabbing the first available V-sync format is the proper way to expose vsync to users. 

Can you elaborate why this is the best approach for exposing v-sync? Please pardon my ignorance if it is something obvious.

Geometror commented:
It is a reimplementation of #36862, but moved to DisplayServer as requested by reduz. In addition I removed the broken vsync via compositor option.
Managing v-sync that way(choosing the first available v-sync present mode) could keep it simple and uniform between Vulkan and the upcoming GL ES renderer. As giarve mentioned VK_PRESENT_MODE_FIFO_KHR in #36862 is the only mode required to be supported, so it makes sense to prefer it over the others as the driver may support it the best. However, I think providing the user with an option to choose the prefered mode would be a good idea. (mainly between MAILBOX[could reduce input lag] and FIFO[guaranteed supported], FIFO_RELAXED isn't that useful imo) 
Do you think that would be a better way to expose vsync to users?

clayjohn commented:
I'm not certain what the best way to expose the vsync settings to the user is. 

I don't think falling back to ``VK_PRESENT_MODE_FIFO_RELAXED_KHR`` makes sense when a user disables v-sync but does not support ``VK_PRESENT_MODE_IMMEDIATE_KHR``.

I think one option may be to have a toggle (between on and off, if ``VK_PRESENT_MODE_IMMEDIATE_KHR`` is not available then try the selected v-sync mode) and a dropdown menu (for v-sync mode, defaulting to ``VK_PRESENT_MODE_FIFO_KHR`` and falling back to ``VK_PRESENT_MODE_FIFO_KHR`` if selected mode is unavailable). 

That being said, I'm not sure that it makes sense to do it like that, for android it looks like ``VK_PRESENT_MODE_SHARED_DEMAND_REFRESH_KHR`` or ``VK_PRESENT_MODE_SHARED_CONTINUOUS_REFRESH_KHR`` may be more suitable when v-sync is disabled (because ``VK_PRESENT_MODE_IMMEDIATE_KHR`` is [never available](https://vulkan.gpuinfo.org/listsurfacepresentmodes.php?platform=android))

Geometror commented:
Ok, I will look into that. By digging through the codebase I found out that it is more work than I initially thought, so I should change this PR to a draft as there are still some things to decide, restructure and implement. Maybe I should ask reduz if he has concrete plans for v-sync atm.

clayjohn commented:
@Geometror that is a good idea. Reduz typically has some ideas for this kind of stuff, and it is important to work on soon as we are approaching alpha.

Geometror commented:
Reimplemented after discussion on IRC with @Calinou, @reduz and @lawnjelly 
The v-sync option is now represented as an enum and implemented on all available platforms, for a summary of all final changes see my edited initial comment (https://github.com/godotengine/godot/pull/48622#issue-638265155)

Could someone test it on Linux and OSX?

Calinou commented:
Testing again on Linux + NVIDIA:

### Disabled

![image](https://user-images.githubusercontent.com/180032/120119780-f173fe80-c199-11eb-9d30-2052c0f63c6e.png)

### On

![image](https://user-images.githubusercontent.com/180032/120119796-051f6500-c19a-11eb-972d-1781566fd1c0.png)

The reported FPS is pretty much always 143-144 FPS.

### Adaptive

![image](https://user-images.githubusercontent.com/180032/120119833-254f2400-c19a-11eb-85eb-bffb236c19ab.png)

Interestingly, the FPS reported is slightly higher than my monitor's refresh rate (144 Hz). It can go up to 148 FPS.

### Mailbox (Triple Buffering)

![image](https://user-images.githubusercontent.com/180032/120119872-54659580-c19a-11eb-9976-9659c280262e.png)

**Note:** When changing the project setting, press <kbd>Ctrl + S</kbd> *twice* to make sure the setting is saved. This is an editor bug unrelated to this PR.

Calinou commented on core/config/project_settings.cpp:
```diff
@@ -1114,8 +1114,8 @@ ProjectSettings::ProjectSettings() {
 
-	// Keep the enum values in sync with the `DisplayServer::ScreenOrientation` enum.
 	custom_prop_info["display/window/handheld/orientation"] = PropertyInfo(Variant::INT, "display/window/handheld/orientation", PROPERTY_HINT_ENUM, "Landscape,Portrait,Reverse Landscape,Reverse Portrait,Sensor Landscape,Sensor Portrait,Sensor");
+	custom_prop_info["display/window/vsync/vsync_mode"] = PropertyInfo(Variant::STRING, "display/window/vsync/vsync_mode", PROPERTY_HINT_ENUM, "disabled,on,adaptive,mailbox/triple buffering");
```
Capitalize every value for consistency with other property hints, and rename "On" to "Enabled" to put it in contrast with "Disabled":

```suggestion
	custom_prop_info["display/window/vsync/vsync_mode"] = PropertyInfo(Variant::STRING, "display/window/vsync/vsync_mode", PROPERTY_HINT_ENUM, "Disabled,Enabled,Adaptive,Mailbox (Triple Buffering)");
```

Geometror replied:
Ok, changed and replaced "Mailbox(Tripple Buffering)" with just "Mailbox" like @lawnjelly suggested as true triple buffering in Vulkan is only achieved with three images in the swapchain, although it has a similar effect. (I forgot to rename that property earlier)

Geometror replied:
Is it fine like this?

Calinou replied:
Looks good to me now.

bruvzg commented:
Tested it on macOS 11.4 with AMD FirePro D300:
* Disabled - runs with unlimited FPS (300-340 FPS).
* All other options - runs at constant FPS equal to double display refresh rate, 120 FPS if display is set to 60 Hz, 60 FPS if it's set to 30 Hz (tested with two different hiDPI and loDPI displays).

Geometror commented:
@bruvzg By looking at the code again it seems like the fallback to the guaranteed supported FIFO present mode works correctly, so your graphics card/driver might not support FIFO_RELAXED or MAILBOX. However, the framerate which is double of the display's refresh rate is weird, likely a driver issue. 

@Calinou With the first 3 modes I got the same results. But in Mailbox mode a basic scene like yours runs close to the framerate with v-sync disabled (around 3000 FPS) on my machine(Windows 10/NVIDIA). That is a little weird, but it could be also a driver issue...

bruvzg commented on servers/display_server.cpp:
```diff
@@ -305,29 +304,13 @@ void DisplayServer::set_icon(const Ref<Image> &p_icon) {
-void DisplayServer::_set_use_vsync(bool p_enable) {
-	WARN_PRINT("VSync not supported by this display server.");
+void DisplayServer::window_set_vsync_mode(DisplayServer::VSyncMode p_vsync_mode, WindowID p_window) {
+	WARN_PRINT("Changing the VSync mod is not supported by this display server.");
```
```suggestion
	WARN_PRINT("Changing the VSync mode is not supported by this display server.");
```

bruvzg commented on servers/display_server.cpp:
```diff
@@ -305,29 +304,13 @@ void DisplayServer::set_icon(const Ref<Image> &p_icon) {
-bool DisplayServer::vsync_is_using_via_compositor() const {
-	return false;
+DisplayServer::VSyncMode DisplayServer::window_get_vsync_mode(WindowID p_window) const {
+	WARN_PRINT("Changing the VSync mod is not supported by this display server.");
```
```suggestion
	WARN_PRINT("Changing the VSync mode is not supported by this display server.");
```

lawnjelly commented:
Just out of interest, and as someone who isn't really familiar with this area :smile: .. is 'adaptive' a better name in some way that just using the vulkan name RELAXED or FIFO_RELAXED?

With a quick google search it seems the name adaptive (apart from being a non-specific term, and seeming to describe anything that isn't on or off) seems as though it could be confusing as it is used for at least 2 different methods:
https://en.wikipedia.org/wiki/FreeSync
https://www.asus.com/support/FAQ/1009271/

> ASUS FreeSync product is certified by both AMD and VESA standard, so it can be called as VESA Adaptive-Sync as well.

FreeSync / VESA Adaptive-Sync seems to be designed to avoid tearing on suitable monitors.

Here nvidia use the word 'adaptive' to describe something closer to FIFO_RELAXED:
https://www.nvidia.com/en-gb/geforce/technologies/adaptive-vsync/

> VK_PRESENT_MODE_FIFO_RELAXED_KHR:
This mode only differs from the previous one if the application is late and the queue was empty at the last vertical blank. Instead of waiting for the next vertical blank, the image is transferred right away when it finally arrives. This may result in visible tearing.

https://vulkan-tutorial.com/Drawing_a_triangle/Presentation/Swap_chain#page_Presentation-mode

bruvzg commented on main/main.cpp:
```diff
@@ -1329,7 +1306,26 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
+	{
+		String vsync_mode = GLOBAL_DEF("display/window/vsync/vsync_mode", "Enabled");
+
+		if (vsync_mode == "Disabled") {
+			print_line("<<Disabled");
+			window_vsync_mode = DisplayServer::VSYNC_DISABLED;
+		} else if (vsync_mode == "Enabled") {
+			print_line("<<On");
+			window_vsync_mode = DisplayServer::VSYNC_ENABLED;
+		} else if (vsync_mode == "Adaptive") {
+			print_line("<<Adaptive");
+			window_vsync_mode = DisplayServer::VSYNC_ADAPTIVE;
+		} else if (vsync_mode == "Mailbox") {
+			print_line("<<Mailbox");
+			window_vsync_mode = DisplayServer::VSYNC_MAILBOX;
+		} else {
+			print_line("v-sync mode unknown.");
+			window_vsync_mode = DisplayServer::VSYNC_ENABLED;
+		}
```
It's probably better to use `print_verbose` (need some code rearrangement, since `verbose` flag is set on the line 1344).

Geometror replied:
Oh, forgot to remove those print_line statements. They were just for debugging, so i removed the first four and replaced the last one with WARN_PRINT.

Geometror commented:
@lawnjelly I understand your point with all these branded synchronization techniques, but I think Adaptive VSync is more clear than FIFO_RELAXED since it behaves similar across rendering APIs (so it would be the same for a possible GLES fallback renderer) and is widely known as "VSync off when framerate falls below screens refresh rate, VSync on if it stays above". However, I updated the documentation so that the user gets a simple summary on how it behaves.

Geometror commented:
Rebased and fixed to compile again after the DisplayServerHeadless addition.

clayjohn commented on doc/classes/DisplayServer.xml:
```diff
@@ -1143,5 +1137,13 @@
 		</constant>
 		<constant name="WINDOW_EVENT_DPI_CHANGE" value="6" enum="WindowEvent">
 		</constant>
+		<constant name="VSYNC_DISABLED" value="0" enum="VSyncMode">
+		</constant>
+		<constant name="VSYNC_ENABLED" value="1" enum="VSyncMode">
+		</constant>
+		<constant name="VSYNC_ADAPTIVE" value="2" enum="VSyncMode">
+		</constant>
+		<constant name="VSYNC_MAILBOX" value="3" enum="VSyncMode">
+		</constant>
```
These should be filled out

Geometror replied:
Done.

clayjohn commented on doc/classes/ProjectSettings.xml:
```diff
@@ -443,7 +443,7 @@
 		<member name="debug/settings/fps/force_fps" type="int" setter="" getter="" default="0">
 			Maximum number of frames per second allowed. The actual number of frames per second may still be below this value if the game is lagging.
-			If [member display/window/vsync/use_vsync] is enabled, it takes precedence and the forced FPS number cannot exceed the monitor's refresh rate.
+			If [member display/window/vsync/vsync_mode] is set to ENABLED or ADAPTIVE, it takes precedence and the forced FPS number cannot exceed the monitor's refresh rate.
```
```suggestion
			If [member display/window/vsync/vsync_mode] is set to [code]ENABLED[/code] or [code]ADAPTIVE[\code], it takes precedence and the forced FPS number cannot exceed the monitor's refresh rate.
```

Geometror replied:
Done, but changed to "[code]Enabled[/code] or [code]Adaptive[/code]" to match with the actual text in the dropdown menu.

clayjohn commented on doc/classes/ProjectSettings.xml:
```diff
@@ -530,12 +530,11 @@
+		<member name="display/window/vsync/vsync_mode" type="String" setter="" getter="" default="&quot;Enabled&quot;">
+			Sets the VSync mode for the main game window.
+			With [code]Enabled[/code] and [code]Mailbox[/code] the image is displayed only on vertical blanking periods, so no tearing is visible.
+			[code]Adaptive[/code] behaves like disabled VSync when the framerate drops below the screen's refresh rate to reduce stuttering(although tearing may be visible), otherwise VSync is enabled to avoid tearing.
```
```suggestion
			[code]Adaptive[/code] behaves like [code]DISABLED[/code] VSync when the framerate drops below the screen's refresh rate to reduce stuttering(although tearing may be visible), otherwise VSync is enabled to avoid tearing.
```

Geometror replied:
Linked the enum destription instead of describing the different modes directly, so that they are at one place.

clayjohn commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -1304,16 +1304,41 @@ Error VulkanContext::_update_swap_chain(Window *window) {
+		window->vsync_mode = DisplayServer::VSYNC_ENABLED; //Set to default
+	}
+
+	print_verbose("Choosen present mode:" + String(string_VkPresentModeKHR(window->presentMode)));
```
```suggestion
	print_verbose("Using present mode: " + String(string_VkPresentModeKHR(window->presentMode)));
```

Geometror replied:
Done.

clayjohn commented on platform/windows/display_server_windows.cpp:
```diff
@@ -1119,7 +1119,7 @@ bool DisplayServerWindows::window_can_draw(WindowID p_window) const {
 	ERR_FAIL_COND_V(!windows.has(p_window), false);
 	const WindowData &wd = windows[p_window];
-	return wd.minimized;
+	return !wd.minimized;
```
This seems unrelated to the rest of the PR. Although the original does look wrong, was this connected to a bug you noticed?

Geometror replied:
Yes, this is a typo and prevented rendering the frames as fast as possible when disabling v-sync on Windows.

akien-mga replied:
This should probably go in its own PR then to be reviewed and merged independently of this one.

Geometror replied:
I could do that, but is that really necessary for this tiny change since it is a bug that prevented the v-sync setting to work properly on Windows?

akien-mga replied:
It looks to me like a bug that would also affect Godot on Windows regardless of this PR, no? 
This method is used in `Window::can_draw()` and in `main.cpp`:
```
OS::get_singleton()->add_frame_delay(DisplayServer::get_singleton()->window_can_draw());
```
So it has potential for changing behavior for Windows users significantly (hopefully for the better), and thus it's better if it's merged independently so that its effect can be assessed independently of this PR for VSync.

Of course you need the fix here so you can also keep the change here and rebase once the other PR is merged (which is likely to come sooner than this here PR given the difference in complexity of the changes).

Geometror replied:
Ok, then I will split this up.

Geometror replied:
Opened  #49253

Geometror commented:
I'm sorry, I have no idea how this got accidently closed... could someone reopen it?

bruvzg commented:
> could someone reopen it?

No, there's `0` commits in PR, that's why it was closed, and it's impossible to reopen it until there are commits. Seems like something gone wrong with the rebase.

clayjohn commented on core/config/project_settings.cpp:
```diff
@@ -1114,8 +1114,8 @@ ProjectSettings::ProjectSettings() {
 
 	_add_builtin_input_map();
 
-	// Keep the enum values in sync with the `DisplayServer::ScreenOrientation` enum.
```
Should this comment be deleted? Seems unrelated to this PR

Geometror replied:
No, sorry, I deleted that comment accidently.

clayjohn commented on doc/classes/DisplayServer.xml:
```diff
@@ -1143,5 +1139,18 @@
+		</constant>
+		<constant name="VSYNC_MAILBOX" value="3" enum="VSyncMode">
+			Displays the most recent image in the queue on vertical blanking intervals, while rendering to the other images (no tearing is visible).
+			Altough not guaranteed, the images can be rendered as fast as possible, which may reduce input lag.
```
```suggestion
			Although not guaranteed, the images can be rendered as fast as possible, which may reduce input lag.
```

Geometror replied:
Oops, fixed.

akien-mga commented on doc/classes/DisplayServer.xml:
```diff
@@ -995,6 +978,19 @@
+			<description>
+				Sets the VSync mode of the given window.
+				See [enum DisplayServer.VSyncMode] for possible values and how they affect the behavior of your application.
+				Depending on the platform and used renderer, the engine will fall back to [code]Enabled[/code], if the desired mode is not supported.
```
I know there was some back and forth on this but it seems @clayjohn wasn't familiar with this convention: you can refer to constants directly with `[constant VSYNC_ENABLED]` to create a hyperlink.

Geometror replied:
Changed.

akien-mga commented on doc/classes/DisplayServer.xml:
```diff
@@ -1143,5 +1139,18 @@
+			Default vertical synchronization mode, the image is displayed only on vertical blanking intervals (no tearing is visible).
+		</constant>
+		<constant name="VSYNC_ADAPTIVE" value="2" enum="VSyncMode">
+			Behaves like [code]VSYNC_DISABLED[/code] when the framerate drops below the screen's refresh rate to reduce stuttering (tearing may be visible), otherwise vertical synchronization is enabled to avoid tearing.
```
This should be `[constant VSYNC_DISABLED]`.

Geometror replied:
Changed.

akien-mga commented on doc/classes/ProjectSettings.xml:
```diff
@@ -530,12 +530,10 @@
+		<member name="display/window/vsync/vsync_mode" type="String" setter="" getter="" default="&quot;Enabled&quot;">
+			Sets the VSync mode for the main game window.
+			See [enum DisplayServer.VSyncMode] for possible values and how they affect the behavior of your application.
+			Depending on the platform and used renderer, the engine will fall back to [code]Enabled[/code], if the desired mode is not supported.
```
Here we could also use `[constant DisplayServer.VSYNC_ENABLED]`, though I think for the project settings there's merit in giving the name of the actual user-facing value in the drop-down list, so it's good as is.

Geometror replied:
I would leave it as is because of the reason you already mentioned.

akien-mga commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -1304,16 +1304,41 @@ Error VulkanContext::_update_swap_chain(Window *window) {
+			break;
+	}
+
+	//Check if the requested mode is available
```
Not super important, but we usually style comments this way (initial space, final dot):
```
// Check if the requested mode is available.
```
Same for other comments.

Geometror replied:
Done.

akien-mga commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -2062,6 +2087,15 @@ String VulkanContext::get_device_pipeline_cache_uuid() const {
 }
 
+DisplayServer::VSyncMode VulkanContext::get_vsync_mode(DisplayServer::WindowID p_window) const {
+	return windows[p_window].vsync_mode;
```
Are we sure that `p_window` will always be a valid index? Otherwise this could crash and would need an `ERR_FAIL_COND` or `ERR_FAIL_INDEX` check to guard it.

Geometror replied:
Done, it throws an error and prints out a message when trying to set/get the VSync mode of a non-existent window.

akien-mga reviewed (commented):
Looks pretty good!

Geometror committed:
Restructure and reimplement vsync options
-Add a v-sync mode setting which allows to choose between DISABLED, ON, ADAPTIVE and MAILBOX
-Removed the V-Sync via Compositor option

Geometror commented:
Rebased.
@Calinou Do you think I should include the editor VSync option from your 4.0 PR (#48364) as part of the whole VSync options reimplementation? Because when/if this is merged, that PR must be redone anyway, so I could save you some work.

clayjohn commented:
@Geometror I think it is best for each PR to do one thing. My preference is to merge this PR first and then look at Calinou's PR after.

Calinou reviewed (approved):
Tested locally with this PR rebased against the latest `master` branch. All options seem to work as expected on Linux + NVIDIA.

**Edit:** Tested on a Windows 10 + Intel IGP setup, it works as expected there too.

Calinou commented:
Thanks!

Calinou commented on core/config/project_settings.cpp:
```diff
@@ -1114,6 +1114,7 @@ ProjectSettings::ProjectSettings() {
 
 	// Keep the enum values in sync with the `DisplayServer::ScreenOrientation` enum.
 	custom_prop_info["display/window/handheld/orientation"] = PropertyInfo(Variant::INT, "display/window/handheld/orientation", PROPERTY_HINT_ENUM, "Landscape,Portrait,Reverse Landscape,Reverse Portrait,Sensor Landscape,Sensor Portrait,Sensor");
+	custom_prop_info["display/window/vsync/vsync_mode"] = PropertyInfo(Variant::STRING, "display/window/vsync/vsync_mode", PROPERTY_HINT_ENUM, "Disabled,Enabled,Adaptive,Mailbox");
```
I just noticed this is a *string* enum rather than an integer enum. I think using an integer enum would be better as it's more Godot-y overall. (We're trying to move away from string usage when possible in 4.0 onwards.)

Geometror replied:
I presume I just copied how it was done for the screen orientation (which was before you changed it to an integer enum). But you are right, an integer enum is cleaner as it wouldn't need that ugly if-elseif construct in main.cpp. I will open a PR soon to change that.

vnen commented on platform/linuxbsd/display_server_x11.h:
```diff
@@ -362,6 +362,9 @@ class DisplayServerX11 : public DisplayServer {
 	virtual void window_set_ime_active(const bool p_active, WindowID p_window = MAIN_WINDOW_ID);
 	virtual void window_set_ime_position(const Point2i &p_pos, WindowID p_window = MAIN_WINDOW_ID);
 
+	virtual void window_set_vsync_mode(DisplayServer::VSyncMode p_vsync_mode, WindowID p_window = MAIN_WINDOW_ID) override;
+	virtual DisplayServer::VSyncMode window_get_vsync_mode(WindowID p_vsync_mode) const override;
```
This class didn't have `override` before this, so now Clang complain about the others that are missing.

Geometror replied:
Is the use of `override` desired? If so, I could add it where it should be (only display_server_iphone and display_server_osx use it currently), otherwise I could open a PR to remove the keyword from all display_server classes except osx and iphone.

vnen replied:
> Is the use of `override` desired?

Yes, it is. I'm not sure why it wasn't used in those classes before.

Geometror replied:
I opened a PR for this: #50513
